### PR TITLE
Fix #85, #2615: Emit `used-before-assignment` in final or except blocks where try statements could have failed

### LINKED ISSUE #85: pylint does not recognize potential undefined variables
Originally reported by: **Mirko Dziadzka (BitBucket: [mirkodziadzka](http://bitbucket.org/mirkodziadzka), GitHub: @mirkodziadzka?)**

---

The following code should generate a warning, that the variable res may be undefined in  the finally clause

```
#!python
try:
    # throw exception here to make res undefined
    res = 42
finally:
    print res
```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/85

### LINKED ISSUE #2615: Not catching use of undefined variables in except clauses
This report is similar to others which mention undetected use of undefined variables after if False or in the finally branch of a try/finally statement. I think use of undefined variables in except clauses is perhaps a higher priority fix. Consider this function:

```
def func(url, headers):
    "doc"
    try:
        resp = requests.get(url, headers)
    except requests.exceptions.SSLError:
        print("SSL error - headers %s", resp.request.headers)
```

It's clear that the variable resp is used while undefined. Missing it is more critical than in most other cases because it is generally going to be executed infrequently. If, as in this case, only one variable is assigned in the try clause and it wasn't defined beforehand, then an undefined-variable message should be emitted. If more than one variable is assigned in the try clause, then a possibly-undefined-variable message should go out over the wire. I'm not suggesting that pylint should be able to do any kind of sophisticated analysis. If it wasn't defined before the try clause, is assigned in the try clause, then referenced in the except clause, it should be flagged.

If the user assigns 37 different variables in the try clause and references fourteen of them in the except clause, the voluminous amount of output will also serve as a good clue that they need to tighten up their try clause (quoting from PEP 8: "Additionally, for all try/except clauses, limit the try clause to the absolute minimum amount of code necessary.") :-)

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

**Previously**
Pylint assumed assignments in the Try block of a Try/Finally or Try/Except pair would succeed. In cases of failure, though, attempting to use such names in a final or except block would raise `UnboundLocalError`.

**Now**
Pylint emits `used-before-assignment` to account for the fact that the statement might have been interrupted by an exception.

Closes #85, Closes #2615

### CODE REVIEW

**Comment by DanielNoord** [comment]
I actually had an interesting discussion about this with one of the maintainers of `pylance`:
https://github.com/microsoft/pylance-release/issues/1958

Basically it comes down to whether we want to assume all `try` statement fail or all succeed. 

Personally I'm inclined to say let's assume it is succeeds. As a user I will be more annoyed by having to add `disable`'s everywhere because one of my tools is producing false positives compared to running into exceptions because my tools are too "dumb". I prefer dumbness over potentially incorrect strictness, but this might be personal preference.

**Comment by jacobtylerwalls** [comment]
Thanks for the speedy reply!

> As a user I will be more annoyed by having to add disable's everywhere because one of my tools is producing false positives

So my suggestion here for the user would be to just define the variables before the try rather than add disables.

> Basically it comes down to whether we want to assume all try statement fail or all succeed.

I agree this is a design question. I think this approach splits the difference well. We assume the try statements fail for the sake of finally statements relying on names that might not exist, which is easy for users to fix by initializing their variables, but we assume the try statements succeed for the sake of evaluating `unused-variable`, which I would agree would be quite annoying to disable if the variable is used later in the try block. How does that strike you?

I know it's not surveying opinions on the other side very well, but there's 13 +1's on the ticket. I tend to agree with the Pylance maintainer that if you're depending on a name in the finally block, then that really is an error--either something should be moved out of the try, or else the variable should be initialized before the try.

**Comment by Pierre-Sassoulas** [comment]
>If you are confident that a statement will never raise an exception, you can move it outside of the `try` block

I found this argument from Eric Traut pretty convincing, honestly. By adding a statement in a Try except the user tell us that the code can fail so we should probably treat the code as possibly failing. If something never fail it should be outside the try and the resulting code will be clearer and better.

**Comment by DanielNoord** [comment]
Does it make sense to create `"possibly-used-before-assignment"` for this? Like we have `"possibly-unused-variable"`? I think @Pierre-Sassoulas and I discussed at some point prior.

If not, I think we should go with the majority opinion and continue with this PR!

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM,! Thank you. I think that taking the same decision than pylance is the safe bet here. But I'm putting this in 2.13 so opinion can be heard about the controversial part of the issue if required.

File: pylint/checkers/variables.py
Comment (jacobtylerwalls) [inline pylint/checkers/variables.py:629]: Sorry for the additional push after approval, but I think this is safer. I noticed while working on the similar issue #4761.

File: ChangeLog
Comment (DanielNoord) [inline ChangeLog:12]: Can you add a comment about this to the description of the message? I have been working on auto generating documentation for each individual message and it would be helpful if the documentation displayed this.

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:623]: Could you use `n.statement(future=True)` here? We will trigger a `DeprecationWarning` otherwise. Same for all the other `statement` calls.

File: tests/functional/u/use/used_before_assignment_issue85.py
Comment (DanielNoord) [inline tests/functional/u/use/used_before_assignment_issue85.py:1]: Could you maybe add a link to the issue here? And perhaps even a docstring explaining that we are assuming the try failed? That might be helpful for future reference.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Small changes here and there! Rest looks good! Glad to see such an old issue finally resolved even if I did not initially agree with the solution. I'm must admit I do think this might be the best solution for this problem going forward. Thanks @jacobtylerwalls for putting this to our attention! 😄

**Comment by jacobtylerwalls** [comment]
Let's not merge until I solve #2615 with a very similar block. I think I can use the same branch.

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:395]: Perfect, thanks!

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:569]: Is this change necessary because of the change to `get_next_to_consume`? I do not know why but the docstring of this method does not seem to indicate that the name not being in `self.to_consume` is an option.

I find the code of the consumption checker to be quite fragile so I'm worried this might break something that was working based on an unwritten assumption.
I don't think this change necessarily breaks anything but I wonder if this function should be called in these cases anyway or if we should have returned in an earlier method that calls `mark_as_consumed`.

Comment (jacobtylerwalls) [reply pylint/checkers/variables.py:569]: > Is this change necessary because of the change to get_next_to_consume? 

Yes, but it will not be necessary after the refactor in #5241 plus the introduction of `consumed_uncertain` in #5402. If you want to hold both PRs for the refactor, that's fine, but if you want to merge this one I will remove the unneeded block after rebasing #5402 on #5241.

The solution in these PRs is that some variables should be partially-consumed: i.e. consumed for the purposes of not being "unused" in the except/finally blocks, or after the try; but still able to emit `unused-before-assignment` if they are used in those blocks (improperly). The old pattern unconditionally marks everything as consumed before exiting the function, so without adding a new attribute like `consumed_uncertain` to store things off, this PR had to let them be marked as consumed twice and added this short-circuit.

Comment (DanielNoord) [reply pylint/checkers/variables.py:569]: Let's do #5241 > #5402 > #5384 then. That seems to make the most sense 😄

I'll have another good look at #5402 after #5241 has been merged. I really want to make the consumption checker as intuitive as possible seeing the large amount of issues and false positives/negatives we currently have. I believe the code being much harder to read compared to other parts of the code are a large reason why these issues stay open. These PRs are really valuable and are a good way to start improving on it, thanks for that!

**Comment by Pierre-Sassoulas** [comment]
@jacobtylerwalls , we merged #5402 (🎉), but I'm not sure how to resolve the conflict, could you rebase on the latest main, please ?

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:408]:
```suggestion
        "Used when a local variable is accessed before its assignment or within "
```

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:668]: Let's store `node.statement` in a local variable. We're calling it 4/5 times in this block.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks! Most of this already looks good.

I thought about `node.statement` previously but never commented about it. Upon re-review I think it makes sense.

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:408]: Let's merge this in the current state and use https://github.com/PyCQA/pylint/pull/5506 to perfect this description 😄

**Review by DanielNoord — APPROVED** [review]

File: pylint/checkers/variables.py
Comment (DanielNoord) [inline pylint/checkers/variables.py:411]: This last part is not actually true for this PR right? Might as well remove it and keep the git history accurate.

**Pull request merged**
